node/src/block_ranges.rs

```diff
@@ -443,6 +443,57 @@ impl Sub<&BlockRanges> for BlockRanges {
     }
 }
 
+impl Sub<RangeInclusive<u64>> for BlockRanges {
+    type Output = Self;
+
+    fn sub(mut self, range: RangeInclusive<u64>) -> Self::Output {
+        self.remove_relaxed(range).expect("shouldn't panic");
+        self
+    }
+}
+
+impl SubAssign<RangeInclusive<u64>> for BlockRanges {
+    fn sub_assign(&mut self, range: RangeInclusive<u64>) {
+        self.remove_relaxed(range).expect("shouldn't panic");
+    }
+}
+
+impl IntoIterator for BlockRanges {
+    type Item = u64;
+
+    type IntoIter = BlockRangesIterator;
+
+    fn into_iter(self) -> Self::IntoIter {
+        let mut ranges = self.0.into_iter();
+        let current_range = ranges.next();
+
+        BlockRangesIterator {
+            ranges,
+            current_range,
+        }
+    }
+}
+
+pub struct BlockRangesIterator {
+    ranges: smallvec::IntoIter<[RangeInclusive<u64>; 2]>,
+    current_range: Option<RangeInclusive<u64>>,
+}
+
+impl Iterator for BlockRangesIterator {
+    type Item = u64;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = self.current_range.as_mut()?.next();
+
+        if next.is_none() {
+            self.current_range = self.ranges.next();
+            return self.current_range.as_mut()?.next();
+        }
+
+        next
+    }
+}
```

Comment:
Add a `pop_tail` (similar to `pop_head`) and then you can implement iterator without having `current_range`. Also why not implementing iterator on `BlockRanges` itself?

Reply:
I think I started from `IntoIterator` and it sort of went that way :P
`impl Iterator for BlockRanges` is with `pop_tail` is indeed simpler, changed :+1: